BlackCubes/foodtrucks-django-react-native
# Foodtruck REST API

## Changes
1. Installed Django REST framework and added it to the settings.
2. Created serializers for the `Product`, `Truck`, and `TruckImage` models.
2. Created views for the `Product` and `Truck` models and serializers.
3. Created routes for the `Product` and `Truck` views.
4. Routed the `foodtruck` app to the main urls.

## Purpose
There needs to be a REST API for the Foodtruck in order for the data to be sent to some client.

## Approach
To fully utilize the REST API implementation, a toolkit called [Django REST framework](https://www.django-rest-framework.org/) was installed. Once installed, it was added to the main settings in order for Django to use it.

In order to convert the SQL models into JSON to achieve a RESTful API, the REST framework utilizes a serializer which converts models and querysets into native Python datatypes that can then be rendered into JSON ([source](https://www.django-rest-framework.org/api-guide/serializers/)).

Three serializers were created for the `Product`, `TruckImage`, and `Truck` models. The `Product` serializer has a relation to the `Truck` model, and with its own properties, it would output a JSON of:

```
{
  "uuid": String,
  "name": String,
  "slug": String,
  "info": String,
  "image": String,
  "price": Number,
  "quantity": Number,
  "is_available": Boolean,
  "truck": Object,
  "reviews": Array,
  "likes": Array
}
```

The `"reviews"` and `"likes"` properties will be implemented later once their serializers are finished, and the `Product` serializer also has a relation to the `Review` and `Like` models.

The `TruckImage` serializer has a relation to the `Truck` model, and with its own properties, it would output a JSON of:

```
{
  "uuid": String,
  "image": String,
  "is_profile_image": Boolean,
  "truck": Object
}
```

The `Truck` serializer has a relation to the `Product` and `TruckImage` models, and with its own properties, it would output a JSON of:

```
{
  "uuid": String,
  "name": String,
  "slug": String,
  "info": String,
  "phone_number": String,
  "email": String,
  "website": String,
  "products": Array,
  "images": Array,
  "events": Array
}
```

The `"events"` property will be implemented later once its serializer is finished, and the `Truck` serializer also has a relation to the `Event` model.

Once these three serializers were finished, views were created for retrieving and listing the `Truck` and `Product` models combined with their serializers. The only method the user can do with this is the `GET` method.

Then, URLs were wired up to the `Truck` and `Product` views, and lastly, the main URL was also wired up to the main `url` file.

## Testing
1. Pull changes.
2. Open the terminal, go to the `/server` folder, and run the virtual environment: `env/Scripts/activate` on Windows, or `env/bin/activate` on Mac.
3. Install the necessary packages from the `requirements.txt` file: `pip install -r requirements.txt`.
4. If you haven't done so already, make the migrations: `python manage.py makemigrations` and then `python manage.py migrate`.
5. Open up the shell: `python manage.py shell`.
6. In the shell, create a dummy foodtruck and product.
7. Quit the shell, then run the server -> `python manage.py runserver`.
8. On your browser, go to `http://localhost:8000/api/v1/foodtrucks/` and make sure your dummy foodtruck is outputted.
9. Now, go to `http://localhost:8000/api/v1/foodtrucks/products` and make sure your dummy product is outputted.
10. Kill the server and flush your database to erase your dummy data (would only remove all data) -> `python manage.py flush`.

## Learning
[Django REST framework](https://www.django-rest-framework.org/).

[Serializers - Django REST framework](https://www.django-rest-framework.org/api-guide/serializers/).

Closes #16
